instance/tests/test_tasks.py

```diff
@@ -121,48 +128,84 @@ def test_deactivate_old_appservers(self, provisioning_succeeds):
         instance = OpenEdXInstanceFactory()
 
-        self.mock_spawn_appserver.return_value = None  # Mock provisioning failure
+        # Disable mocking of retry-enabled spawn_appserver
+        self.spawn_appserver_patcher.stop()
+        self.addCleanup(self.spawn_appserver_patcher.start)
+
+        # Mock provisioning failure
+        mock_spawn.return_value = None
+
         tasks.spawn_appserver(instance.ref.pk, num_attempts=3, mark_active_on_success=True)
 
-        self.assertEqual(self.mock_spawn_appserver.call_count, 3)
+        # Check mocked functions call count
+        self.assertEqual(mock_spawn.call_count, 3)
         self.assertEqual(self.mock_make_appserver_active.call_count, 0)
 
+        # Confirm logs
         self.assertTrue(any("Spawning new AppServer, attempt 1 of 3" in log.text for log in instance.log_entries))
         self.assertTrue(any("Spawning new AppServer, attempt 2 of 3" in log.text for log in instance.log_entries))
         self.assertTrue(any("Spawning new AppServer, attempt 3 of 3" in log.text for log in instance.log_entries))
 
-    def test_num_attempts_successful(self):
+    @patch('instance.models.openedx_instance.OpenEdXInstance._spawn_appserver')
+    @patch('instance.models.openedx_appserver.OpenEdXAppServer.provision')
+    def test_num_attempts_successful(self, mock_provision, mock_spawn):
         """
         Test that if num_attempts > 1, the spawn_appserver task will stop trying to provision
         after a successful attempt.
         """
         instance = OpenEdXInstanceFactory()
 
-        self.mock_spawn_appserver.return_value = 10  # Mock provisioning failure
+        # Disable mocking of retry-enabled spawn_appserver
+        self.spawn_appserver_patcher.stop()
+        self.addCleanup(self.spawn_appserver_patcher.start)
+
+        # Mock successful provisioning
+        mock_provision.return_value = True
+        mock_spawn.return_value = make_test_appserver(instance)
+
         tasks.spawn_appserver(instance.ref.pk, num_attempts=3, mark_active_on_success=True)
 
-        self.assertEqual(self.mock_spawn_appserver.call_count, 1)
+        # Check mocked functions call count
+        self.assertEqual(mock_spawn.call_count, 1)
+        self.assertEqual(mock_provision.call_count, 1)
         self.assertEqual(self.mock_make_appserver_active.call_count, 1)
 
+        # Confirm logs
         self.assertTrue(any("Spawning new AppServer, attempt 1 of 3" in log.text for log in instance.log_entries))
         self.assertFalse(any("Spawning new AppServer, attempt 2 of 3" in log.text for log in instance.log_entries))
         self.assertFalse(any("Spawning new AppServer, attempt 3 of 3" in log.text for log in instance.log_entries))
 
-    def test_one_attempt_default(self):
+    @patch('instance.models.openedx_instance.OpenEdXInstance._spawn_appserver')
+    @patch('instance.models.openedx_appserver.OpenEdXAppServer.provision')
+    def test_one_attempt_default(self, mock_provision, mock_spawn):
```

Comment:
Also test in case the first provision fails, there is should still only be 1 attempt.

Reply:
👍